consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java

```diff
@@ -21,70 +21,457 @@
+      throw new RatisRequestFailedException(e);
+    }
+    return reply;
+  }
 
   @Override
-  public void start() {}
+  public void start() {
+    // TODO awkward, should we add IOException to method signature?
+    try {
+      server.start();
+    } catch (IOException ignored) {
+      try {
+        server.close();
+      } catch (IOException ignored1) {
+      }
+    }
+  }
 
   @Override
-  public void stop() {}
+  public void stop() {
+    try {
+      server.close();
+    } catch (IOException ignored) {
+    }
+  }
 
   @Override
   public ConsensusWriteResponse write(
       ConsensusGroupId groupId, IConsensusRequest IConsensusRequest) {
-    return ConsensusWriteResponse.newBuilder().build();
+
+    RaftClient client = clientMap.get(Utils.toRatisGroupId(groupId));
+    TSStatus writeResult = null;
+    try {
+      ByteBufferConsensusRequest request = (ByteBufferConsensusRequest) IConsensusRequest;
+      Message message = Message.valueOf(ByteString.copyFrom(request.getContent()));
+      RaftClientReply reply = client.io().send(message);
+      writeResult = Utils.deserializeFrom(reply.getMessage().getContent().asReadOnlyByteBuffer());
+    } catch (IOException | TException e) {
+      return ConsensusWriteResponse.newBuilder()
+          .setException(new RatisRequestFailedException(e))
+          .build();
+    }
+    return ConsensusWriteResponse.newBuilder().setStatus(writeResult).build();
   }
 
+  /** Read directly from LOCAL COPY notice: May read stale data (not linearizable) */
   @Override
   public ConsensusReadResponse read(ConsensusGroupId groupId, IConsensusRequest IConsensusRequest) {
-    return ConsensusReadResponse.newBuilder().build();
+
+    RaftGroup group = raftGroupMap.get(Utils.toRatisGroupId(groupId));
+    if (group == null || !group.getPeers().contains(myself)) {
+      return ConsensusReadResponse.newBuilder()
+          .setException(new ConsensusGroupNotExistException(groupId))
+          .build();
+    }
+
+    RaftClientReply reply = null;
+    try {
+      assert IConsensusRequest instanceof ByteBufferConsensusRequest;
+      ByteBufferConsensusRequest request = (ByteBufferConsensusRequest) IConsensusRequest;
+
+      RaftClientRequest clientRequest =
+          RaftClientRequest.newBuilder()
+              .setServerId(server.getId())
+              .setClientId(localFakeId)
+              .setGroupId(Utils.toRatisGroupId(groupId))
+              .setCallId(localFakeCallId.incrementAndGet())
+              .setMessage(Message.valueOf(ByteString.copyFrom(request.getContent())))
+              .setType(RaftClientRequest.staleReadRequestType(0))
+              .build();
+
+      reply = server.submitClientRequest(clientRequest);
+    } catch (IOException e) {
+      return ConsensusReadResponse.newBuilder()
+          .setException(new RatisRequestFailedException(e))
+          .build();
+    }
+
+    Message ret = reply.getMessage();
+    assert ret instanceof ReadLocalMessage;
```

Comment:
Use if statement and throw an exception~

Reply:
Thx, I'll use if & throw at IStateMachine. Here, I'm sure _ret_ must be an instance of ReadLocalMessage, since _ret_ is only returned from ApplicationStateMachineProxy.query. I put assert here to ensure this. If this assert fails, it MUST be a program bug.